### ISSUE TITLE
Drawing isuue and Feature request

### ISSUE THREAD

**Issue by gitsamhub** [pubpub-zz/ppInk#92]
@pubpub-zz Thank you for suggesting this wonderful tool. 

As I mentioned before as an artist I use it only for drawing on reference images. And it is pretty good at that but with some issues.

**Bugs :**
1. When drawing with pen it gives me this weird problem of stroke flickering or something. Please check attached video.
2. Also, when i try to move the strokes it moves with custom transparent background. Kindly check video @0:26 frame

**Feature Request:**
Since I use graphic tablet to draw I just need facility to assign three shortcut keys:
1. Key to increase and decrease pen width (e.g. + and - key)
2. Key to pop up color picker 
3. Key to take screenshot of just the drawing without desktop screen. (e.g. PNG file)

https://user-images.githubusercontent.com/79096837/117650051-db33dd80-b1ad-11eb-84fb-d7eddd66a215.mp4

Thanks.

**Comment by pubpub-zz** [comment]
Happy you'fd enjoying.
About the points you've raised:

> **Bugs :**
>     1. When drawing with pen it gives me this weird problem of stroke flickering or something. Please check attached video.

I've activated the FitToCurve property that provide nicer curver but which are inducing the jumps. I will think about a way to disable that in a nice and open manner
>     2. Also, when i try to move the strokes it moves with custom transparent background. Kindly check video @0:26 frame

There is 2 moving options the original pan that moves everything 
![pan_act](https://user-images.githubusercontent.com/4083478/117657637-16c4ac80-b19a-11eb-9a49-f78a6d904bd3.png)
and the new onve that moves only one:
![pan1_act](https://user-images.githubusercontent.com/4083478/117657719-322fb780-b19a-11eb-8fa0-f941cd9cda13.png)
The second shows a box around the object to move and has a red cursor (your cursor is black at 0.28)

> **Feature Request:**
> Since I use graphic tablet to draw I just need facility to assign three shortcut keys:
> 
>     1. Key to increase and decrease pen width (e.g. + and - key)

Will think about it. And alternative/complement I'm currrently working on is to propose an REST http interface that would provides a mean to control ppInk from another program device. is such a solution compatible with your graphic tablet?
> 
>     2. Key to pop up color picker

Have you tried the option with the rotating pen selection shortcut (cf #69 - sorry forgot to add the help in main page about this option

> 
>     3. Key to take screenshot of just the drawing without desktop screen. (e.g. PNG file)

Not sure about feasability will try

**Comment by gitsamhub** [comment]
Thank you for addressing. 

For now I can deal with every other issue and can find workaround for those. But flickering drawing is the major issue that just doesn't fit my specific workflow.

I love the simplicity of geovens (gInk). Specially the smooth and precise drawing function. But missed absence of keyboard shortcuts. 

So only thing I am looking forward is nice and clean drawing pen without any artifacts in ppInk.

Looking forward to next update.

Thanks.

**Comment by pubpub-zz** [comment]
yes

**Comment by gitsamhub** [comment]
@pubpub-zz Hi, so far the new changes are working nicely. 

Just a little thing I wanted to ask. Since  we can not export only drawing as PNG file I found a workaround is to change the background to pure white and then take snapshot. After that import that image in another application and then delete the white area. However, for that i will have to go to option settings every time to change background to white whenever i finish my drawing, which is quite tedious. So is it possible to assign shortcut key to change the background (or even better cycle through it)? If is it too much of a work then just shortcut for opening options settings will work too.

Again, I really appreciate your time and concern regarding developing this tool. Thanks.

**Comment by gitsamhub** [comment]
Hi, I am not sure why did it lag but now after restarting my system it is working as expected.

**Few things:**

1. I noticed that when i take snapshot after opening the app, the first snapshot comes with transparent background. But after that it works fine( means just captures the drawing as intended).

2. Regarding choosing color. When I change pen with different color my brush width also changes because it is set as such in settings. Is it possible to just change the color of the pen and keep its width as it is which i am using? Because it is already possible to change width with keyboard shortcut, user can change it as required. Or this can be solved with just assigning shortcut for popping up color setting. That way i can choose desired wide variety of color shades instead of few pen icons in the UI panel.
Basically I just need a keyboard shortcut to pop up color setting for which right now i have to right click on pen icon every time.

Otherwise I can see this tool is improving quite nicely. Thanks

**Comment by pubpub-zz** [comment]
> * I noticed that when i take snapshot after opening the app, the first snapshot comes with transparent background. But after that it works fine( means just captures the drawing as intended).

Sorry I do not understand what you mean exactly. can you give an example.
 
> * Regarding choosing color. When I change pen with different color my brush width also changes because it is set as such in settings. Is it possible to just change the color of the pen and keep its width as it is which i am using? Because it is already possible to change width with keyboard shortcut, user can change it as required. Or this can be solved with just assigning shortcut for popping up color setting. That way i can choose desired wide variety of color shades instead of few pen icons in the UI panel.
>   Basically I just need a keyboard shortcut to pop up color setting for which right now i have to right click on pen icon every time.


I've introduced a solution : long press (>3s) on the pen to modified shortcut will open the edit pen dialog

> * Now I might be nitpicking here, but i was curious to know how the width sensitivity of brush is controlled in this app? e.g. Maximum stroke width and minimum stroke width. I checked my graphic tablet settings but it can only make pressure firm and smooth. Right now brush can produce maximum width nicely but is it possible to make minimum width much thinner? I hope that makes sense.

Sorry The pressure is directly implement in .NET framework : I can not adjust the min/max pressure

I've also find a solution to prevent the outlining. also the capture stroke only option is now adjustable in the options
[ppInk64_12a.zip](https://github.com/pubpub-zz/ppInk/files/6468829/ppInk64_12a.zip)

**Comment by gitsamhub** [comment]
Hi,

1. Black pixel outlining issue is solved for fully opaque strokes. However if stroke is transparent then it looks pixelated (jagged) around the edges.

2. I have set customized transparent background in the settings. 
I open the app, then draw with pen, then take snapshot. If i check the snapshot image, it includes transparent background in it. Please check Snapshot_1 image (It includes background in it).
![Snapshot_1](https://user-images.githubusercontent.com/79096837/118091555-947df780-b3e8-11eb-810f-949812983087.png)

I draw again, then take snapshot. This time it doesn't include transparent background in it. 
Please check Snapshot_2 image (It doesn't include background).
![Snapshot_2](https://user-images.githubusercontent.com/79096837/118092937-5b468700-b3ea-11eb-8cb5-ce17a8f05313.png)

3. When i take snapshot, app minimizes and clears my drawing automatically. (I want it to stay as it is without clearing the drawing).

4. My screen display is 2560 x 1440 pixels. And i have scaled it up 125% for better visibility. I noticed that color picker doesn't pick up exact location in this setting.

5. Lets say i draw three strokes on the screen. So when click (P) to pan them it moves one selected stroke. If I click (p) again, it duplicates the selected stroke. If I click (p) again it moves all strokes including customized transparent background I have set in the settings. (clicking 'P' for third time should only move all strokes excluding transparent background)

6. Regarding opening up color adjustment setting. I am aware of the press and hold pen function with it. However i have disabled it in my system because it causes problems in other places while using system. (I just want the ability to set a keyboard shortcut that will open color settings).

I am just providing feedback here while using it for my specific workflow, and not really complaining about the functionality. Nevertheless, this tool is incredible and has a lot of potential and a lot of people can use it for their work.

Thank you.

**Comment by pubpub-zz** [comment]
>     1. Black pixel outlining issue is solved for fully opaque strokes. However if stroke is transparent then it looks pixelated (jagged) around the edges.

Can you show me an example, Mine looks good:
![2021-05-13 20-46-13](https://user-images.githubusercontent.com/4083478/118172442-aa151080-b42c-11eb-9395-6d3fad00610a.png)

>     2. I have set customized transparent background in the settings.

Actually there was a bug : the background is not applied after exiting through a snapshot.
By the way, remember the background is a filed rectangle : You can delete it  moving the eraser on the screen border and then click.

>     3. When i take snapshot, app minimizes and clears my drawing automatically. (I want it to stay as it is without clearing the drawing).

You can do this with a long press of the snapshot button, or alt+HK_Snapshot.
Remember also that if you have closed your inking inadvertently, you can restore it with the load button ( which is loading the autosave.strokes.txt file)

>     4. My screen display is 2560 x 1440 pixels. And i have scaled it up 125% for better visibility. I noticed that color picker doesn't pick up exact location in this setting.

I've not been able to reproduce can you provide snapshot or video ?

>     3. Lets say i draw three strokes on the screen. So when click (P) to pan them it moves one selected stroke. If I click (p) again, it duplicates the selected stroke. If I click (p) again it moves all strokes including customized transparent background I have set in the settings. (clicking 'P' for third time should only move all strokes excluding transparent background)

The background is currently a filled rectangle, it is therefore move with all the others during pan process. will check feasability to prevent it from moving
 
>     4. Regarding opening up color adjustment setting. I am aware of the press and hold pen function with it. However i have disabled it in my system because it causes problems in other places while using system. (I just want the ability to set a keyboard shortcut that will open color settings).

first, sorry, I just seen that the pen edition on long HK press was not coded for rotating color. now implemented in the beta attached

I have some problem to understand : with the following configuration
![image](https://user-images.githubusercontent.com/4083478/118181665-4b559400-b438-11eb-9df1-f38790db97e9.png)
you can press and release 1 to move to next pen and if you press and hold 1, you will edit next color.

> I am just providing feedback here while using it for my specific workflow, and not really complaining about the functionality. Nevertheless, this tool is incredible and has a lot of potential and a lot of people can use it for their work.

**Comment by pubpub-zz** [comment]
New beta fixing subtool panel appearing wrongly, and prevent background moving during pan
[ppInk64_13b.zip](https://github.com/pubpub-zz/ppInk/files/6475159/ppInk64_13b.zip)

**Comment by gitsamhub** [comment]
1. Since i can delete transparent rectangle with eraser key that issue is solved now.
2. Also with Alt+Hotkey for snapshot is also working fine. I didn't know about that hotkey.

3.Color Picker issue: 
![Image_1](https://user-images.githubusercontent.com/79096837/118223422-e1240a00-b49e-11eb-8122-81c3050b3bb7.jpg)
![Image_2](https://user-images.githubusercontent.com/79096837/118223430-e4b79100-b49e-11eb-8d0e-5f9da0a4cbca.jpg)
![color picker](https://user-images.githubusercontent.com/79096837/118223435-e7b28180-b49e-11eb-9e64-e33386212095.jpg)
-As you can see in image_1 with blue rectangle I am referring to scale and layout of Windows 10.
-In Image_2 you can see I have scaled my screen display  to 125%.
-In Color Picker Image I am referring to this little icon that picks up color from any pixel on the active screen area.
Problem is when i click and drag this to choose color from the screen It doesn't pick up color from selected area precisely, because my display size is scaled up to 125%. However it works if Display is scaled up to100%. 

4. Is it possible to assign a Hotkey for 'Color Picker'?

5. Is it possible to assign a Hotkey for just to open 'Pen Adjustment' box?
( Right now i can do that by pressing and holding pen for some time and box appears. This function is interesting, however, this is not really efficient because every time user has to wait for that box to pop up. 
I would really like to suggest just A SINGLE HOTKEY that will open this box. This is much needed functionality.)

6. 
> you can press and release 1 to move to next pen and if you press and hold 1, you will edit next color.
Cycling through pen with a Hotkey is good idea but It creates other issues. If i assign Hotkey 1 for it I can not use other number Hotkey e.g. 2,3,4,5..etc for other pens. Also pressing and holding it changes the color of next pen. I don't see any advantage in having this functionality. 
However you can set 'Tab' as Hotkey to cycle through different pens. But at the same time we can also have ability to select specific pen by pressing numbers like 2, 3, 4.. etc.

Thank you.

**Comment by NOVAinc** [comment]
I have noticed that in the last few beta versions posted around these topics, disabling smooth curves results in jagged edges on strokes. I believe version 10a still retained antialiasing even with smooth curves off. It would be good having that back. I don't know if it's a regression from the Capture Strokes Only feature, but I tested with that option on, and using smooth curves, and the transparent pictures did retain anti aliasing.

**Comment by pubpub-zz** [comment]
@NOVAinc,
Since version 10a, capability to disable smoothing/FitToCurve was introduced, maybe Anti-aliased was not forced on. I just did it to ensure.

@gitsamhub 
>     2\. Also with Alt+Hotkey for snapshot is also working fine. I didn't know about that hotkey.

Good. I will complete readme


> 3.Color Picker issue:
You have to adjust PPP parameters (sorry for french)
![image](https://user-images.githubusercontent.com/4083478/118373666-5cc5aa00-b5b8-11eb-852c-e46946b666d2.png)
![image](https://user-images.githubusercontent.com/4083478/118373713-8252b380-b5b8-11eb-82cb-a23595a5a8dd.png)

>4. Is it possible to assign a Hotkey for 'Color Picker'?
I have to completely code this function (ca not re-use a framework) Will push it into the stack

**Comment by gitsamhub** [comment]
Hi, thank you for constantly improving this tool.
Yes, Dpi scaling issue is solved now.

> I have to completely code this function (ca not re-use a framework) Will push it into the stack

I really appreciate your time and concern for focusing on this complex functionalities. This would really make it very user friendly.

**Comment by NOVAinc** [comment]
I thought I might just comment here since the jaggy edges was brought on this topic. I think I've found out when it is smooth and when it's not. I've tested version 17b and there seems to be no antialiasing for translucent strokes. I've captured this solid gray stroke next to a orange highlighter, and in the orange stroke edges can be clearly seen pixelated.
![2021-05-17 06-18-16](https://user-images.githubusercontent.com/19545604/118465151-fb94f800-b6d7-11eb-98b7-bf7d0ba428f5.png)

Hope this helps!

**Comment by pubpub-zz** [comment]
On my side, I can see the antialiasing on both colors:
![image](https://user-images.githubusercontent.com/4083478/118483562-711fb900-b716-11eb-84c7-edad3f87b26f.png)
(capture zoomed with gimp)
There is not difference in the code with the transparency(set to 0 if no transparency...)
With the smoothing off, I can not see any difference:
![image](https://user-images.githubusercontent.com/4083478/118484227-48e48a00-b717-11eb-8775-80e6a404af6e.png)
The input (mouse ou touchscreen) does not have any impact neither...
Do you see any difference between stylus and mouse ?

**Comment by NOVAinc** [comment]
I have tested further, here is a picture I labelled and it seems to be affect translucent strokes made with pen only. Mouse strokes get properly anti-aliased both solid and translucent. Pen strokes get anti-aliased using solid pens only. I have tested this by increasing stroke width and using the built-in pens.
![strokes](https://user-images.githubusercontent.com/19545604/118553287-80aefa00-b736-11eb-81ab-fd08df3b8cb8.png)

**Comment by pubpub-zz** [comment]
@NOVAinc
It seems to be due to the way Microsoft have coded the library 😕 ( strokes are directly drawn using Microsoft.Ink...) there is nothing I can do about

**Comment by pubpub-zz** [comment]
Hi @gitsamhub,
I've just finished an alpha version of the color picker.
2 options will allow to engage the mode : key shortcut(by default no shortcut), or longclick on penwidth button
When engaging the mode, the transparency of the current pen is used, but you can modify it with the mouse wheel, a click on the icon or a new press of the shortcut will exit the mode, but will be able to select color/transparency and apply them to current pen with the click.
There is still some bugs to fix (interactions with selection of other pens/tools/pointer mode / not an arrow cursor when mouse over toolbar/missing REST API) : do not hesitate to report what you will see about.

[ppInk64_21a.zip](https://github.com/pubpub-zz/ppInk/files/6524465/ppInk64_21a.zip)

**Comment by gitsamhub** [comment]
Hi @pubpub-zz ,

- New color picker is working as expected. Thank you for developing and doing that extra work.

- Screenshot problem: Although I have checked 'Snapshot strokes only' in general setting, when i take snapshot it includes background in it. Of course I can erase background by side clicking on screen with eraser tool but that is not efficient way if I take frequent snapshots of my drawings.

- For the background I have set custom background color (Pure white color with 50% alpha). But that setting keeps going back to default. Actually, this happened in previous version as well.

*Just for the reference I have attached images of my Hotkey and general settings that i use for my specific workflow. 
I was also curious about what 'Snapshot in Pointer Mode' setting does?

![General setting](https://user-images.githubusercontent.com/79096837/119219259-3e086b80-bb02-11eb-98ee-be7a5f7a5b17.png)

![My Hotkey  Setting](https://user-images.githubusercontent.com/79096837/119219255-3cd73e80-bb02-11eb-95b5-a204b4610e28.png)

Thank you

**Comment by pubpub-zz** [comment]
>     * New color picker is working as expected. Thank you for developing and doing that extra work.
 
>     * Screenshot problem: Although I have checked 'Snapshot strokes only' in general setting, when i take snapshot it includes background in it. Of course I can erase background by side clicking on screen with eraser tool but that is not efficient way if I take frequent snapshots of my drawings.

The background is removed now when capturing stroke only

>     * For the background I have set custom background color (Pure white color with 50% alpha). But that setting keeps going back to default. Actually, this happened in previous version as well.

Found : the config.ini parameter label is wrong : gray_Board1
 
>     * Request: Do you think if is it still possible to assign hotkey for popping up 'Pen Adjustment' window. Right now it appears after right clicking on pen icons. This particular facility i am requesting only for the purpose of choosing different shades of color which i use frequently. (I am not aware of the technical aspect of implementing that function though).

With the new pickup system, it was easier to implement: Added

> *Just for the reference I have attached images of my Hotkey and general settings that i use for my specific workflow.

Can you provide also the Config.ini and Hotkeys.ini ; it will be easier to deploy/test. Thx

> I was also curious about what 'Snapshot in Pointer Mode' setting does?

This is a specific snapshot solution proposed to document solutions with menus but not only : have a look at #65
 
> Thank you

**Comment by pubpub-zz** [comment]
oups forgot the new beta:
The points I've raised in the very first alpha are normally all solved:
[ppInk64_22a.zip](https://github.com/pubpub-zz/ppInk/files/6526288/ppInk64_22a.zip)

**Comment by gitsamhub** [comment]
Hi @pubpub-zz 

- 'Snapshot strokes only' option working now.
- Custom background kept going to default setting. However, I changed that setting in Config.ini and now it is fixed.
- Color pickup hotkey working nicely.

-Can i suggest if possible to implement this thing. It would be really useful if you change the current color wheel to similar with below attached images. I guess it is called 'HSV Square'. It is common in almost all painting and coloring applications. This is efficient and convenient way for any user to choose color shades. I am more concerned about the square part in the middle which gives ability to choose dark and light values of a single color. (Highly required function)

![Choose color option_1](https://user-images.githubusercontent.com/79096837/119230678-f520d900-bb3a-11eb-96fe-6f1e79254610.png)
![Choose color option_2](https://user-images.githubusercontent.com/79096837/119230682-f6520600-bb3a-11eb-901e-744ab0b5855f.png)

-I noticed that current color wheel has somewhat sluggish response to adjusting color slider. E.g. when i open color wheel and try to move vertical triangular slider, there is slight delay to it. It doesn't move until i move slider little far away. Means, it doesn't respond to slightly adjusted distance. (Again this might not be issue with program itself but it is noticeable, especially while using graphic tablet)

-Regarding selection of strokes: When i make a drawing It consist of couple of strokes. Lets say I want to move this drawing I can use hotkey 'P'. After that i make another drawing with couple of strokes, now if i want to move it, it moves everything on screen including the first drawing that i have made before. This behavior is obvious though. But it can be avoided if there is facility to select multiple strokes at a time. Right now only one stroke can be selected at a time which is not really helpful, because drawing is made of more than one stroke. So if possible 'drag and select' function is required which is similar to when we select multiple folders on desktop.

-And one last thing, Can we have Hotkey for exiting program completely. Or 'Right click functionality' Or 'Hotkey' for desktop icon of ppink that will open context menu options. From there user can click on: Options, About and Exit.

I am also attaching 'Config.ini' and 'Hotkeys.ini' file with it.

[Config and hotkey file.zip](https://github.com/pubpub-zz/ppInk/files/6526710/Config.and.hotkey.file.zip)

Thank you very much for bearing with me. :)

**Comment by pubpub-zz** [comment]
>     * 'Snapshot strokes only' option working now.
>     * Custom background kept going to default setting. However, I changed that setting in Config.ini and now it is fixed.> 
>     * Color pickup hotkey working nicely.

All good 😊

> -Can i suggest if possible to implement this thing. It would be really useful if you change the current color wheel to similar with below attached images. I guess it is called 'HSV Square'. It is common in almost all painting and coloring applications. This is efficient and convenient way for any user to choose color shades. I am more concerned about the square part in the middle which gives ability to choose dark and light values of a single color. (Highly required function)

I rely on a COTS library https://github.com/cyotek/Cyotek.Windows.Forms.ColorPicker. They are currently not proposing any view like the one you are looking for. If you have proposal for another GPL library, I can have a look. For the moment, I propose to create a new issue for tracking, but no immediate solution is forecasted

> -I noticed that current color wheel has somewhat sluggish response to adjusting color slider. E.g. when i open color wheel and try to move vertical triangular slider, there is slight delay to it. It doesn't move until i move slider little far away. Means, it doesn't respond to slightly adjusted distance. (Again this might not be issue with program itself but it is noticeable, especially while using graphic tablet)

Can you do a video ?

> -Regarding selection of strokes: When i make a drawing It consist of couple of strokes. Lets say I want to move this drawing I can use hotkey 'P'. After that i make another drawing with couple of strokes, now if i want to move it, it moves everything on screen including the first drawing that i have made before. This behavior is obvious though. But it can be avoided if there is facility to select multiple strokes at a time. Right now only one stroke can be selected at a time which is not really helpful, because drawing is made of more than one stroke. So if possible 'drag and select' function is required which is similar to when we select multiple folders on desktop.

I had other requests about selection of multiple strokes. This will be for next release

> -And one last thing, Can we have Hotkey for exiting program completely. Or 'Right click functionality' Or 'Hotkey' for desktop icon of ppink that will open context menu options. From there user can click on: Options, About and Exit.

Can you clarify the use case of exiting program hotkey : personnaly I Start ppInk at begginning of my session and keep it in background for ever.
I personnally think that the context menu is sufficient. I currently use the rightclick on the floating window to move it and I want to keep it for this for use with tablets/touchscreen in an easy way (right click= long press on touch screens)

**Comment by gitsamhub** [comment]
`Can you do a video ?`

https://user-images.githubusercontent.com/79096837/119234608-7b91e680-bb4c-11eb-9276-40eb215893ef.mp4
I am not sure if you can see it in real time, but if you go frame by frame you can clearly see, when i drag slider down it doesn't move until i travel some distance.
```

Can you clarify the use case of exiting program hotkey : personnaly I Start ppInk at begginning of my session and keep it in background for ever.
I personnally think that the context menu is sufficient. I currently use the rightclick on the floating window to move it and I want to keep it for this for use with tablets/touchscreen in an easy way (right click= long press on touch screens)
```
I agree with you. Actually i did not notice small icon in my taskbar that i can use for context options. So it is solved now.

**Comment by pubpub-zz** [comment]
I Propose to close this thread now

**Comment by gitsamhub** [comment]
> 
> 
> I Propose to close this thread now

Thank you for making wonderful tool. Excited to try new features.
